### PR TITLE
Fix Add/Remove WIP on pull request title failure

### LINKED ISSUE #29997: Can't remove WIP from PR title
### Description

When creating a PR in WIP state it is not possible (1.21.9) to remove the WIP: prefix from title.

Steps to reproduce:

- Create a repo / or use an existing one without CODEOWNERs files
- Create a PR with WIP: Prefix in title
- Open the PR  and try to remove the WIP Prefix

https://try.gitea.io/sebastian-sauer/test-wip-bug/pulls/1

Log output:
```
2024/03/22 09:57:56 ...rs/web/repo/issue.go:2163:UpdateIssueTitle() [E] ChangeTitle: exit status 128 - fatal: ambiguous argument '': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
 - fatal: ambiguous argument '': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

### Gitea Version

1.21.9

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Binary from download site

### Database

None

### PR DESCRIPTION
**Pull request by lunny**
Fix #29997 

--------
`pr.HeadCommitID` cannot be used directly, it needs to be loaded after pr is loaded from the database.

### CODE REVIEW

**Review by wolfogre — APPROVED** [review]

File: tests/integration/pull_review_test.go [hunk 1]
```diff
@@ -89,6 +90,23 @@ func TestPullView_CodeOwner(t *testing.T) {
...
+			prURL := fmt.Sprintf("/user2/test_codeowner/pulls/%d", pr.Index)
+			req := NewRequest(t, "GET", prURL)
+			resp := session.MakeRequest(t, req, http.StatusOK)
+			htmlDoc := NewHTMLParser(t, resp.Body)
+
+			req = NewRequestWithValues(t, "POST", prURL+"/title", map[string]string{
+				"_csrf": htmlDoc.GetCSRF(),
+				"title": "[WIP] Test Pull Request",
+			})
+			session.MakeRequest(t, req, http.StatusOK)
```
Comment (wxiaoguang) [inline tests/integration/pull_review_test.go]: I think it should also test whether the PR is really updated.

In many cases, panic occurs in `http.StatusOK` .

Comment (lunny) [reply tests/integration/pull_review_test.go]: 6d42421

**Review by wxiaoguang — DISMISSED** [review]

File: tests/integration/pull_review_test.go [hunk 2]
```diff
@@ -89,6 +90,29 @@ func TestPullView_CodeOwner(t *testing.T) {
...
+			assert.EqualValues(t, "[WIP] Test Pull Request", prUpdated1.Issue.Title)
+
+			req = NewRequestWithValues(t, "POST", prURL+"/title", map[string]string{
+				"_csrf": htmlDoc.GetCSRF(),
+				"title": "Test Pull Request",
+			})
+			session.MakeRequest(t, req, http.StatusOK)
+			prUpdated2 := unittest.AssertExistsAndLoadBean(t, &issues_model.PullRequest{BaseRepoID: repo.ID, HeadBranch: "codeowner-basebranch"})
+			assert.NoError(t, prUpdated2.LoadIssue(db.DefaultContext))
+			assert.EqualValues(t, "Test Pull Request", prUpdated2.Issue.Title)
```
Comment (wxiaoguang) [inline tests/integration/pull_review_test.go]: You see, test failed.

```
         pull_review_test.go:115: 
            	Error Trace:	/home/runner/work/gitea/gitea/tests/integration/pull_review_test.go:115
            	Error:      	Not equal: 
            	            	expected: "Test Pull Request"
            	            	actual  : "[WIP] Test Pull Request"
```

File: services/issue/pull.go [hunk 3]
```diff
@@ -90,7 +90,7 @@ func PullRequestCodeOwnersReview(ctx context.Context, pull *issues_model.Issue,
 
 	// https://github.com/go-gitea/gitea/issues/29763, we need to get the files changed
 	// between the merge base and the head commit but not the base branch and the head commit
-	changedFiles, err := repo.GetFilesChangedBetween(mergeBase, pr.HeadCommitID)
+	changedFiles, err := repo.GetFilesChangedBetween(mergeBase, pr.GetGitRefName())
```
Comment (wxiaoguang) [inline services/issue/pull.go:93]: I am not sure whether it is right for a forked repo.

If there are too many unclear things, I would suggest to ignore some non-important errors (only panic in dev mode), don't block 1.21 users.

Comment (wxiaoguang) [reply services/issue/pull.go:93]: See my comment https://github.com/go-gitea/gitea/issues/29997#issuecomment-2015040562

Anything wrong in `PullRequestCodeOwnersReview` shouldn't block user from changing the title.

**Comment by Gr3q** [comment]
Workaround with the API, replace everything in curly brackets.

```bash
curl -X 'PATCH' \
  'https://{host}/api/v1/repos/{user}/{repo}/issues/{pr_index}' \
  -H 'accept: application/json' \
  -H 'authorization: {your auth}' \
  -H 'Content-Type: application/json' \
  -d '{
  "title": "{title}"
}'
```

**Comment by silverwind** [comment]
@lunny please fix tests.

File: services/issue/pull.go (hunk 3 above)
Comment (lunny) [reply services/issue/pull.go:93]: For a forked repository, it's still right. The file `refs/pull/<id>/head` stored the pull request's head commit id. It's right even the branch has been removed.

Comment (wxiaoguang) [reply services/issue/pull.go:93]: But what if it doesn't exist, or something else reports other errors? I do not think any error in PullRequestCodeOwnersReview should block user's other operations like changing the title.

I prefer the error to be ignored but not reporting a 500 error page.

Comment (lunny) [reply services/issue/pull.go:93]: 02833022b2569e9970b2f6dff22682f8308ea94f

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**